### PR TITLE
home page, follow

### PR DESCRIPTION
**Pull request by PhamVanTung**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -7,7 +7,7 @@ def create
     user = User.find_by email: params[:session][:email].downcase
     if user && user.authenticate(params[:session][:password])
```
Comment (ngocvnn) [inline app/controllers/sessions_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -47,6 +48,20 @@ def destroy
     redirect_to users_url
   end
 
+  def following
+    @title = "Following"
+    @user  = User.find params[:id]
+    @users = @user.following.paginate(page: params[:page])
```
Comment (ngocvnn) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -47,6 +48,20 @@ def destroy
...
+    render 'show_follow'
+  end
+
+  def followers
+    @title = "Followers"
+    @user  = User.find params[:id]
+    @users = @user.followers.paginate(page: params[:page])
```
Comment (ngocvnn) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -54,16 +69,8 @@ def user_params
                                  :password_confirmation)
```
Comment (ngocvnn) [inline app/controllers/users_controller.rb]: ??

File: app/models/relationship.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+class Relationship < ActiveRecord::Base
+	belongs_to :follower, class_name: "User"
```
Comment (ngocvnn) [inline app/models/relationship.rb]: indent

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (PhamVanTung) [inline app/controllers/users_controller.rb]: khong bo dc ()

File: app/controllers/sessions_controller.rb (hunk 1 above)
Comment (PhamVanTung) [inline app/controllers/sessions_controller.rb]: khong duoc bo ()

File: app/views/relationships/create.js.erb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+$("#follow_form").html("<%= escape_javascript(render('users/unfollow')) %>")
+$("#followers").html('<%= @user.followers.count %>')
```
Comment (vanlinha3k35pbc) [inline app/views/relationships/create.js.erb]: Add newline at end

File: app/views/relationships/destroy.js.erb [hunk 7]
```diff
@@ -0,0 +1,2 @@
+$("#follow_form").html("<%= escape_javascript(render('users/follow')) %>")
+$("#followers").html('<%= @user.followers.count %>')
```
Comment (vanlinha3k35pbc) [inline app/views/relationships/destroy.js.erb]: Add newline at end

File: app/views/shared/_stats.html.erb [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+    </strong>
+    following
+  </a>
+  <a href="<%= followers_user_path(@user) %>">
+    <strong id="followers" class="stat">
+      <%= @user.followers.count %>
+    </strong>
+    followers
+  </a>
+</div>
```
Comment (vanlinha3k35pbc) [inline app/views/shared/_stats.html.erb]: Add newline at end

File: app/views/shared/_user_info.html.erb [hunk 9]
```diff
@@ -0,0 +1,3 @@
+<h1>	
+	<%= current_user.name %>
+</h1>
```
Comment (vanlinha3k35pbc) [inline app/views/shared/_user_info.html.erb]: Add newline at end

File: app/views/static_pages/home.html.erb [hunk 10]
```diff
@@ -1,5 +1,19 @@
...
+	</aside>
+	<div class="col-md-8">
+		<div class="col-md-4">
+			<%= link_to "Word", '#' , class: "btn btn-lg btn-primary" %>
+		</div>
+		<div class="col-md-4">
+ 	 	<%= link_to "Category", categories_path , class: "btn btn-lg btn-primary" %>
+ 	 </div>
+ 	</div>
 </div>
```
Comment (vanlinha3k35pbc) [inline app/views/static_pages/home.html.erb]: Add newline at end

**Comment by ngocvnn** [comment]
ok

File: app/controllers/relationships_controller.rb [hunk 11]
```diff
@@ -0,0 +1,22 @@
+class RelationshipsController < ApplicationController
+
+	 before_action :logged_in_user
```
Comment (dieunb) [inline app/controllers/relationships_controller.rb]: indent

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -1,6 +1,7 @@
 class UsersController < ApplicationController
 
-  before_action :logged_in_user, except: [:new, :create]
+  before_action :logged_in_user, only: [:index, :edit, :update, :destroy,
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: `except` instead of `only`

File: app/controllers/users_controller.rb [hunk 13]
```diff
@@ -47,6 +48,20 @@ def destroy
     redirect_to users_url
   end
 
+  def following
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: RESTful

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -47,6 +48,20 @@ def destroy
...
+    @title = "Following"
+    @user  = User.find params[:id]
+    @users = @user.following.paginate page: params[:page]
+    render 'show_follow'
+  end
+
+  def followers
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: RESTFul

File: app/views/relationships/create.js.erb [hunk 15]
```diff
@@ -0,0 +1,2 @@
+$("#follow_form").html("<%= escape_javascript(render('users/unfollow')) %>")
```
Comment (dieunb) [inline app/views/relationships/create.js.erb]: `escape_javascript render('users/unfollow')`

File: app/views/relationships/destroy.js.erb [hunk 16]
```diff
@@ -0,0 +1,2 @@
+$("#follow_form").html("<%= escape_javascript(render('users/follow')) %>")
```
Comment (dieunb) [inline app/views/relationships/destroy.js.erb]:
```
escape_javascript render('users/follow')
```

File: app/views/shared/_user_info.html.erb [hunk 17]
```diff
@@ -0,0 +1,3 @@
+<h1>	
+	<%= current_user.name %>
```
Comment (dieunb) [inline app/views/shared/_user_info.html.erb]: indent

File: app/views/static_pages/home.html.erb [hunk 18]
```diff
@@ -1,5 +1,19 @@
 <% provide :title, "Home" %>
-<div class="center jumbotron">
-  <h1>Welcome to the Framgia E-Learning</h1>
-  <%= link_to "Sign up now!", signup_path, class: "btn btn-lg btn-primary" %>
+<div class="row">
+	<aside class="col-md-4">
```
Comment (dieunb) [inline app/views/static_pages/home.html.erb]: indent

File: app/views/static_pages/home.html.erb [hunk 19]
```diff
@@ -1,5 +1,19 @@
...
+     <section class="user_info">
+       <%= render 'shared/user_info' %>
+     </section>
+     <section class="stats">
+        <%= render 'shared/stats' %>
+      </section>
+	</aside>
+	<div class="col-md-8">
+		<div class="col-md-4">
```
Comment (dieunb) [inline app/views/static_pages/home.html.erb]: indent

File: app/views/users/index.html.erb [hunk 20]
```diff
@@ -2,5 +2,5 @@
 <h1>All users</h1>
 
 <ul class="users">
-  <%= render @users %>
+	<%= render @users %>
```
Comment (dieunb) [inline app/views/users/index.html.erb]: indent

File: config/routes.rb [hunk 21]
```diff
@@ -1,16 +1,23 @@
...
   get    'login'   => 'sessions#new'
   post   'login'   => 'sessions#create'
   delete 'logout'  => 'sessions#destroy'
-  resources :users
+
+  resources :categories
+  resources :relationships, only: [:create, :destroy]
+  resources :users do
+    member do
+      get :following, :followers
```
Comment (dieunb) [inline config/routes.rb]: remove

File: app/controllers/followings_controller.rb [hunk 22]
```diff
@@ -0,0 +1,9 @@
+class FollowingsController < ApplicationController
+
+  def index
+    @title = "Following"
+    @user  = User.find params[:user_id]
+    @users = @user.following.paginate page: params[:page]
+
```
Comment (ngocvnn) [inline app/controllers/followings_controller.rb]: remove

File: app/controllers/relationships_controller.rb [hunk 23]
```diff
@@ -0,0 +1,22 @@
+class RelationshipsController < ApplicationController
+
+   before_action :logged_in_user
```
Comment (ngocvnn) [inline app/controllers/relationships_controller.rb]: indent

Comment (ngocvnn) [inline app/controllers/relationships_controller.rb]: indent

File: app/controllers/sessions_controller.rb [hunk 24]
```diff
@@ -7,7 +7,7 @@ def create
...
       log_in user
-      params[:session][:remember_me] == '1' ? remember(user) : forget(user)
+      params[:session][:remember_me] == "1" ? remember(user) : forget(user)
       redirect_back_or user
     else
       flash.now[:danger] = 'Invalid email/password combination'
```
Comment (ngocvnn) [inline app/controllers/sessions_controller.rb]: "

File: app/views/followings/index.html.erb [hunk 25]
```diff
@@ -0,0 +1 @@
+<%= render 'shared/show_follow'%>
```
Comment (ngocvnn) [inline app/views/followings/index.html.erb]: add new line

File: app/views/relationships/create.js.erb [hunk 26]
```diff
@@ -0,0 +1,2 @@
+$("#follow_form").html("<%= escape_javascript render('users/unfollow') %>")
+$("#followers").html('<%= @user.followers.count %>')
```
Comment (ngocvnn) [inline app/views/relationships/create.js.erb]: add new line

File: app/views/relationships/destroy.js.erb [hunk 27]
```diff
@@ -0,0 +1,2 @@
+$("#follow_form").html("<%= escape_javascript render('users/follow') %>")
+$("#followers").html('<%= @user.followers.count %>')
```
Comment (ngocvnn) [inline app/views/relationships/destroy.js.erb]: add new line

File: app/views/shared/_show_follow.html.erb [hunk 28]
```diff
@@ -0,0 +1,21 @@
...
+  <div class="col-md-8">
+    <h3><%= @title %></h3>
+    <% if @users.any? %>
+      <ul class="users follow">
+        <%= render @users %>
+      </ul>
+      <%= will_paginate @users%>
+    <% end %>
+  </div>
+</div>
```
Comment (ngocvnn) [inline app/views/shared/_show_follow.html.erb]: add new line

File: app/views/shared/_stats.html.erb [hunk 29]
```diff
@@ -0,0 +1,15 @@
...
+    </strong>
+    following
+  </a>
+  <a href="<%= user_followers_path(@user) %>">
+    <strong id="followers" class="stat">
+      <%= @user.followers.count %>
+    </strong>
+    followers
+  </a>
+</div>
```
Comment (ngocvnn) [inline app/views/shared/_stats.html.erb]: add new line

File: app/views/shared/_user_info.html.erb [hunk 30]
```diff
@@ -0,0 +1,3 @@
+<h1>  
+  <%= current_user.name %>
+</h1>
```
Comment (ngocvnn) [inline app/views/shared/_user_info.html.erb]: add new line

File: app/views/static_pages/home.html.erb [hunk 31]
```diff
@@ -1,5 +1,19 @@
...
+  </aside>
+  <div class="col-md-8">
+    <div class="col-md-4">
+      <%= link_to "Word", '#' , class: "btn btn-lg btn-primary" %>
+    </div>
+    <div class="col-md-4">
+    <%= link_to "Category", categories_path , class: "btn btn-lg btn-primary" %>
+   </div>
+  </div>
 </div>
```
Comment (ngocvnn) [inline app/views/static_pages/home.html.erb]: add new line

File: app/views/static_pages/home.html.erb [hunk 32]
```diff
@@ -1,5 +1,19 @@
...
+     <section class="stats">
+        <%= render 'shared/stats' %>
+      </section>
+  </aside>
+  <div class="col-md-8">
+    <div class="col-md-4">
+      <%= link_to "Word", '#' , class: "btn btn-lg btn-primary" %>
+    </div>
+    <div class="col-md-4">
+    <%= link_to "Category", categories_path , class: "btn btn-lg btn-primary" %>
```
Comment (ngocvnn) [inline app/views/static_pages/home.html.erb]: indent

File: app/views/users/_follow_form.html.erb [hunk 33]
```diff
@@ -0,0 +1,9 @@
+<% unless current_user?@user %>
```
Comment (ngocvnn) [inline app/views/users/_follow_form.html.erb]: add space

File: app/views/users/_follow_form.html.erb [hunk 34]
```diff
@@ -0,0 +1,9 @@
...
+  <div id="follow_form">
+  <% if current_user.following?@user %>
```
Comment (ngocvnn) [inline app/views/users/_follow_form.html.erb]: add any space

Comment (ngocvnn) [inline app/views/users/_follow_form.html.erb]: indent

File: app/controllers/users_controller.rb [hunk 35]
```diff
@@ -54,16 +54,8 @@ def user_params
...
-      store_location
-      flash[:danger] = "Please log in."
-      redirect_to login_url
-    end
-  end
-
   def correct_user
-    @user = User.find_by params[:id]
+    @user = User.find params[:id]
     redirect_to root_url unless current_user?@user
```
Comment (vanlinha3k35pbc) [inline app/controllers/users_controller.rb]: missing space

File: app/views/shared/_show_follow.html.erb [hunk 36]
```diff
@@ -0,0 +1,21 @@
...
+      <%= render 'shared/stats' %>
+    </section>
+  </aside>
+  <div class="col-md-8">
+    <h3><%= @title %></h3>
+    <% if @users.any? %>
+      <ul class="users follow">
+        <%= render @users %>
+      </ul>
+      <%= will_paginate @users%>
```
Comment (vanlinha3k35pbc) [inline app/views/shared/_show_follow.html.erb]: missing space

File: app/views/static_pages/home.html.erb [hunk 37]
```diff
@@ -1,5 +1,19 @@
...
-  <h1>Welcome to the Framgia E-Learning</h1>
-  <%= link_to "Sign up now!", signup_path, class: "btn btn-lg btn-primary" %>
+<div class="row">
+  <aside class="col-md-4">
+     <section class="user_info">
+       <%= render 'shared/user_info' %>
+     </section>
+     <section class="stats">
+        <%= render 'shared/stats' %>
+      </section>
```
Comment (vanlinha3k35pbc) [inline app/views/static_pages/home.html.erb]: indent

File: app/views/users/_follow_form.html.erb (hunk 33 above)
Comment (vanlinha3k35pbc) [inline app/views/users/_follow_form.html.erb]: miss space

File: app/views/users/_follow_form.html.erb (hunk 34 above)
Comment (vanlinha3k35pbc) [inline app/views/users/_follow_form.html.erb]: miss space

**Comment by ngocvnn** [comment]
ok

File: app/controllers/sessions_controller.rb [hunk 38]
```diff
@@ -7,10 +7,10 @@ def create
     user = User.find_by email: params[:session][:email].downcase
     if user && user.authenticate(params[:session][:password])
```
Comment (vanlinha3k35pbc) [inline app/controllers/sessions_controller.rb]: ()

File: app/views/shared/_stats.html.erb [hunk 39]
```diff
@@ -0,0 +1,15 @@
+<% @user ||= current_user %>
+<div class="stats">
+  <a href="<%= user_followings_path(@user)  %>">
```
Comment (vanlinha3k35pbc) [inline app/views/shared/_stats.html.erb]: ()

File: app/views/shared/_stats.html.erb [hunk 40]
```diff
@@ -0,0 +1,15 @@
...
+    <strong id="following" class="stat">
+      <%= @user.following.count %>
+    </strong>
+    following
+  </a>
+  <a href="<%= user_followers_path(@user) %>">
```
Comment (vanlinha3k35pbc) [inline app/views/shared/_stats.html.erb]: ()

File: app/views/users/_follow.html.erb [hunk 41]
```diff
@@ -0,0 +1,4 @@
+<%= form_for(current_user.active_relationships.build, remote: true) do |f| %>
```
Comment (vanlinha3k35pbc) [inline app/views/users/_follow.html.erb]: ()

File: app/views/users/_follow_form.html.erb [hunk 42]
```diff
@@ -0,0 +1,9 @@
+<% unless current_user? @user %>
+  <div id="follow_form">
+  	<% if current_user.following? @user %>
```
Comment (vanlinha3k35pbc) [inline app/views/users/_follow_form.html.erb]: indent

File: app/views/users/_unfollow.html.erb [hunk 43]
```diff
@@ -0,0 +1,5 @@
+<%= form_for(current_user.active_relationships.find_by(followed_id: @user.id),
+             html: {method: :delete} ,
```
Comment (vanlinha3k35pbc) [inline app/views/users/_unfollow.html.erb]: indent

File: app/views/users/_user.html.erb [hunk 44]
```diff
@@ -4,4 +4,4 @@
     <%= link_to "delete", user, method: :delete,
                                 data: {confirm: "You sure?"} %>
```
Comment (vanlinha3k35pbc) [inline app/views/users/_user.html.erb]: indent

**Pull request merged**